IA-3365 user history backend

quang-le opened the pull request:
Explain what problem this PR is resolving

Related JIRA tickets : IA-3365

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [x] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
- Tell us where the doc can be found (docs folder, wiki, in the code...)

## Changes
- update `Modification` model:
    - use SET_NULL i.o DO_NOTHING for user foreign key
    - Add SQL migration to make existing logs consistent
- Add `ProfileAuditLogger, subclassing `AuditLogger`. The shape of the `new_value` and `past_value` entries matches the shape of existing logs for users and adds a few fields. It was kept flat for ease of use with existing front-end components.
- Add `ProfileAuditSerializer` and related nested serializers:
    - In the `new_value` field, we log a boolean `password_updated`.
    - On deletion, we log the time in a `deleted_at` field, in anticipation of soft delete feature (so the logs will be consistent)
- Add audit logging to 'create`method of `ProfileViewset `
- Refactor `partial_update` method of `ProfileViewset` to raise all potential exceptions before writing to the DB. Add audit logging when update is successful
- Refactor `update_single_profile_from_bulk` in the same way. Add  audit log on profile if update successful. Add audit log for teams if teams are updated
- Add audit logging to `BulkCreateUserFromCsvViewSet`.
- Add fixtures and tests


## How to test
- Login as super user to have full access to the `/api/logs/`
Go to iaso>admin>users
- create a single user
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`
- The latest log should be the user you created. It should have an empty `past_value`. `new_value.password_updated` should be `True`

- update existing user
- repeat above steps and check that values are correct

- select several users in the table
- With the speed dial, select update
- Update all possible fields and save
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`: you should find a log per user
- Go to `/api/logs/contentType=iaso.team&fields=past_value,new_value/`: you should find a log per user and per team modified

- click create from file button and confirm
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`: you should find a log per user


## Notes

I haven't refactored the code for user bulk create because we will go through it again for performance and move it in to a task shortly, so it can be done correctly then

quang-le committed:
CHORE: remove unused methods

- in iaso/utils/__init__.py:
    - as_soup
    - parse_xml_file

quang-le committed:
IA-3365: move profile API files together, update imports

quang-le committed:
IA-3365: add audit log for regular profile API
- add logging for create and partial_update

quang-le committed:
cleanup print

quang-le committed:
IA-3365: add logging to profile bulk update

- Improve profle logging by using ProfileLogger
- Log team changes

quang-le committed:
IA-3365: add logging to profile bulk create

quang-le committed:
IA-3365: add logging for profile deletion (hard delete)

quang-le committed:
IA-3365: fix bad var assignbment

quang-le committed:
IA-3365: improve profile logging
- differentiate source by user agent
- differentiate bulk create from bulk update
- log user role id as well as name

quang-le committed:
IA-3365: log whether password has been updated

quang-le committed:
IA-3365: log id + name for project and org unit

quang-le committed:
IA-3365: add tests for create, update, delete

quang-le committed:
IA-3365: refactor profile partial_update
- Regroup validation to raise errors before writing to the DB to avoid multiple DB saves that are hard to log

quang-le committed:
IA-3365: add test fro profile/me

quang-le committed:
black formatting

quang-le committed:
Merge branch 'main' into IA-3365_user_history_backend

quang-le committed:
IA-3365: add test for bulk create user

quang-le committed:
IA-3365: add permission check on projects for profile update

- Add check on user admin permission to be consistent with bulk update API

quang-le committed:
IA-3365: raise errors before writing DB for bulk update

quang-le committed:
IA-3365: add test for audit logs in bulk update

quang-le committed:
IA-3365: add comments in test code

quang-le committed:
IA-3365: update Modification model

- use SET_NULL i.o DO_NOTHING for user foreign key
- Add SQL migration to make existing logs consistent

hakifran commented on iaso/api/profiles/audit.py:
```diff
@@ -0,0 +1,91 @@
+    class Meta:
+        model = User
+        # We're serializing the password to be able to determine if it has been changed. but we remove them before saving the log itself
+        # DO NOT reuse this serializer if you're going to send the payload to the front-end
```
Why not declare it inline inside the ProfileAuditSerializer to make sure it won't be used any where else

quang-le replied:
Good catch! I should've thought of that, thanks ! 🙏

quang-le committed:
IA-3365: log deletion time

- Add log on field "deleted_at". Set it to null by default as long as no soft delete is implemented
- on hard delete, log a new_value field with deletion time. This should ensure smooth transition for the front-end
- update tests

quang-le committed:
black

quang-le committed:
IA-3365: align log object shape with historic logs

- for easier backwards compatibility when displaying in the front-end

quang-le committed:
IA-3365: chnage profile log to have flat structure

quang-le committed:
IA-3365: add test on pwd and pk

hakifran commented on iaso/api/profiles/profiles.py:
```diff
@@ -360,55 +362,153 @@ def partial_update(self, request, pk=None):
+        if projects is not None:
+            if not request_user.has_perm(permission.USERS_ADMIN):
+                raise PermissionDenied(
+                    f"User with permission {permission.USERS_MANAGED} cannot changed project attributions"
```
**Connot change**

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -32,21 +39,41 @@ def update_single_profile_from_bulk(
+    org_units_to_be_added = []
+    org_units_to_be_removed = []
+    # Raise if necessary and prepare values to be updated
+    if teams_id_added is not None:
```
For me it could be more clear if this condition is combine in one condition with the one at 120th line

quang-le replied:
The thing is I want to raise all exceptions before writing to the DB to avoid having partial DB updates and no logs for them. That's why the code is split

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -32,21 +39,41 @@ def update_single_profile_from_bulk(
+    if teams_id_added is not None:
+        if not user.has_perm(permission.TEAMS):
+            raise PermissionDenied(f"User without the permission {permission.TEAMS} cannot add users to team")
+    if teams_id_removed is not None:
```
same here it could be more clear if this condition is combine in one condition with the one at 132 line

quang-le replied:
Same as above

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
-                )
-            org_unit = OrgUnit.objects.get(pk=location_id)
+    if len(roles_to_be_added) > 0:
+        for role in roles_to_be_added:
```
As there is no any specific logic to each added user role I think profile.iaso_profile.add(*roles_to_be_added) can work well and avoid the for loop.

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
+        for role in roles_to_be_added:
+            role.iaso_profile.add(profile)
+    if len(roles_to_be_removed) > 0:
+        for role in roles_to_be_removed:
```
Same here  profile.iaso_profile.remove(*roles_to_be_removed)

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
+    if len(roles_to_be_removed) > 0:
+        for role in roles_to_be_removed:
+            role.iaso_profile.remove(profile)
+    if len(projects_to_be_added) > 0:
```
Same here

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
+    if len(projects_to_be_added) > 0:
+        for project in projects_to_be_added:
+            project.iaso_profile.add(profile)
+    if len(projects_to_be_removed) > 0:
```
Same here

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
+    if len(projects_to_be_removed) > 0:
+        for project in projects_to_be_removed:
+            project.iaso_profile.remove(profile)
+    if len(org_units_to_be_added) > 0:
```
Same here

hakifran commented on iaso/tasks/profiles_bulk_update.py:
```diff
@@ -65,56 +93,80 @@ def update_single_profile_from_bulk(
-                )
-            org_unit = OrgUnit.objects.get(pk=location_id)
+    if len(org_units_to_be_removed) > 0:
+        for org_unit in org_units_to_be_removed:
```
Same here

quang-le committed:
Merge branch 'main' into IA-3365_user_history_backend

quang-le commented:
Failing Python tests come from the hotfix on csv user creation performance (on main)

hakifran commented:
> Failing Python tests come from the hotfix on csv user creation performance (on main)

@madewulf I think this is related to this https://github.com/BLSQ/iaso/pull/1607 and I reported but I thought this was related to my local configuration

quang-le committed:
Merge branch 'main' into IA-3365_user_history_backend

quang-le committed:
IA-3365: fix tests

quang-le committed:
Merge branch 'main' into IA-3365_user_history_backend

quang-le committed:
IA-3365: merge migrations

hakifran reviewed (changes requested):
@quang-le when the save button is triggered without any changes in any input it create a history where the difference is only "password_updated": false, is it the way it should work? for instance if I add the user to a team there is history created on the user with "password_updated": false

hakifran commented on iaso/api/profiles/profiles.py:
```diff
@@ -372,55 +374,153 @@ def partial_update(self, request, pk=None):
+            # Prevent from username change with existing username
+            raise ProfileError(field="user_name", detail=_("Nom d'utilisateur existant"))
+
+    def validate_user_permissions(self, request, current_account):
```
another question is that why not filter on permission like this permission = get_object_or_404(Permission, codename__startswith="iaso_", codename=permission_codename) just to let know that a given permission not allowed to be attributed or there is a mistake.

quang-le replied:
I don't know I just moved the existing code in another function

hakifran commented on iaso/api/profiles/audit.py:
Permissions are not needed here to be able to log the user changes?

quang-le replied:
No. It's all happening internally, while using other APIs that already control for permissions

hakifran replied:
> No. It's all happening internally, while using other APIs that already control for permissions

You are right!!

quang-le commented:
> @quang-le when the save button is triggered without any changes in any input it create a history where the difference is only "password_updated": false, is it the way it should work? for instance if I add the user to a team there is history created on the user with "password_updated": false

This is a front-end problem with the "User" interface. You shouldn't be able to save when you made no changes

quang-le committed:
Merge branch 'main' into IA-3365_user_history_backend

quang-le committed:
IA-3365: delete duplicate migration

quang-le committed:
IA-3365 code review

- Move ProfileAuditFieldsSerializer
- remove for loops in profile bulk update
- fix typo

quang-le committed:
IA-3365: typo

hakifran reviewed (commented):
@quang-le ![Screenshot from 2024-09-13 10-39-02](https://github.com/user-attachments/assets/bd7f623d-c5c4-4263-a206-77bee0fcd6c6)

mathvdh committed:
Merge branch 'main' of github.com:BLSQ/iaso into IA-3365_user_history_backend

mathvdh committed:
IA-3365 Fix migrations conflicts

mathvdh committed:
IA-3365 fix formatting
